### PR TITLE
Z image lora training

### PR DESCRIPTION
**Pull request by linoytsaban**
add lora training script for https://huggingface.co/Tongyi-MAI/Z-Image

### CODE REVIEW

**Comment by linoytsaban** [comment]
@bot /style

**Comment by sayakpaul** [comment]
Thanks for the script, do you have an example training run as well?

**Comment by linoytsaban** [comment]
running a testing train run with this dataset: https://huggingface.co/datasets/Norod78/Yarn-art-style
https://wandb.ai/linoy/dreambooth-z-image-lora/runs/yqd6pnzg?nw=nwuserlinoy
seems to be learning as expected
left/top: after 18 steps, right/bottom: after 1000 steps
`a panda drinking matcha, yarn art style`
media_images_validation_18_7b60e21cb9e7acc524a0
media_images_test_1000_d48f997c1155d8400f5d

File: examples/dreambooth/README_z_image.md [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+We've added aspect ratio bucketing support which allows training on images with different aspect ratios without cropping them to a single square resolution. This technique helps preserve the original composition of training images and can improve training efficiency.
+
+To enable aspect ratio bucketing, pass `--aspect_ratio_buckets` argument with a semicolon-separated list of height,width pairs, such as:
+
+```bash
+--aspect_ratio_buckets="672,1568;688,1504;720,1456;752,1392;800,1328;832,1248;880,1184;944,1104;1024,1024;1104,944;1184,880;1248,832;1328,800;1392,752;1456,720;1504,688;1568,672"
+```
+
+### Bilingual Prompts
```
Comment (sayakpaul) [inline examples/dreambooth/README_z_image.md:306]: This is stunning stuff!

File: examples/dreambooth/train_dreambooth_lora_z_image.py [hunk 2]
```diff
@@ -0,0 +1,1912 @@
...
+pipeline = AutoPipelineForText2Image.from_pretrained("Tongyi-MAI/Z-Image", torch_dtype=torch.bfloat16).to('cuda')
+pipeline.load_lora_weights('{repo_id}', weight_name='pytorch_lora_weights.safetensors')
+image = pipeline('{validation_prompt if validation_prompt else instance_prompt}').images[0]
+```
+
+For more details, including weighting, merging and fusing LoRAs, check the [documentation on loading LoRAs in diffusers](https://huggingface.co/docs/diffusers/main/en/using-diffusers/loading_adapters)
+
+## License
+
+Please adhere to the licensing terms as described [here](https://huggingface.co/black-forest-labs/FLUX.2/blob/main/LICENSE.md).
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_z_image.py]: Does it share the same licensing as Flux.1?

File: examples/dreambooth/train_dreambooth_lora_z_image.py [hunk 3]
```diff
@@ -0,0 +1,1912 @@
...
+
+                model_pred_list = transformer(
+                    noisy_model_input_list,
+                    timestep_normalized,
+                    prompt_embeds,  # This is a List[torch.Tensor] for Z-Image
+                    return_dict=False,
+                )[0]
+                model_pred = torch.stack(model_pred_list, dim=0)  # (B, C, 1, H, W)
+                model_pred = model_pred.squeeze(2)  # (B, C, H, W)
+                model_pred = -model_pred  # z-Image negates the prediction
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_z_image.py:1706]: Oh interesting. Is this from the pipeline implementation for reference?

File: examples/dreambooth/train_dreambooth_lora_z_image.py [hunk 4]
```diff
@@ -0,0 +1,1912 @@
...
+
+                # these weighting schemes use a uniform timestep sampling
+                # and instead post-weight the loss
+                weighting = compute_loss_weighting_for_sd3(weighting_scheme=args.weighting_scheme, sigmas=sigmas)
+
+                # flow matching loss
+                target = noise - model_input
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_z_image.py:1713]: Should the target be also flipped because of the negation above?

File: examples/dreambooth/train_dreambooth_lora_z_image.py (hunk 2 above)
Comment (linoytsaban) [reply examples/dreambooth/train_dreambooth_lora_z_image.py]: my bad, fixed

File: examples/dreambooth/train_dreambooth_lora_z_image.py (hunk 3 above)
Comment (linoytsaban) [reply examples/dreambooth/train_dreambooth_lora_z_image.py:1706]: yep: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/z_image/pipeline_z_image.py#L553-L558

**Comment by linoytsaban** [comment]
another example, trained on poli's tarot card dataset https://huggingface.co/datasets/multimodalart/1920-raider-waite-tarot-public-domain. 
left: w/o the LoRA, right: w/ the LoRA

image
image

**Comment by sayakpaul** [comment]
The RHS perfectly fits my job description.

**Comment by linoytsaban** [comment]
@sayakpaul last failing test seems unrelated?

**Comment by innokria** [comment]
How to use this script , any demo in HF space will be helpful , so all Lora  training scripts sits under example folder , use case for this should be in the DOC section :)

**Comment by innokria** [comment]
pip install -r requirements_z_image.txt   this file "requirements_z_image.txt " is missing in the folder 

Anyway ZeroGPU failed now will try in L40S
https://huggingface.co/spaces/rahul7star/ZIMG_LORAS
